### PR TITLE
Add Initial Support for ARM MTE Instructions to Rasm2

### PR DESCRIPTION
**Pull request by nathansrf**
- [x] Mark this if you consider it ready to merge
- [x] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**

This PR adds new instructions to the AArch64 assembler to enable encoding of ARM MTE instructions.

I have added support for the following core instructions:
| Instruction | Description             | Format                                                                                                                                                 |
| ----------- | ----------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------ |
| ADDG        | Add with tag            | ADDG <Xd/SP>, <Xn/SP>, #<uimm6>, #<uimm4>                                                                                                              |
| CMPP        | Compare tag             | CMPP <Xn/SP>, <Xm/SP>                                                                                                                                  |
| GMI         | Tag mask insert         | GMI <Xd>, <Xn/SP>, <Xm>                                                                                                                                |
| IRG         | Insert random tag       | IRG <Xd\SP>, <Xn\SP>\{, <Xm>\}                                                                                                                         |
| STG         | Store allocation tag    | STG <Xt/SP>, [<Xn/SP>], \#<simm> (post-index) <br> STG <Xt\SP>, [<Xn\SP>, #<simm>]! (pre-index) <br> STG <Xt\SP>, [<Xn\SP>{, #<simm>}] (signed offset) |
| STGM        | Store multiple (Vector) | STGM <Xt>, [<Xn\SP>]                                                                                                                                   |
| LDG         | Load allocation tag     | LDG <Xt>, [<Xn/SP>{, #<simm>}]                                                                                                                         |
| LDGM        | Load multiple tags      | LDGM <Xt>, [<Xn\SP>]                                                                                                                                   |

### CODE REVIEW

File: libr/arch/p/arm/armass64.c [hunk 1]
```diff
@@ -84,9 +88,12 @@ typedef struct Opcode_t {
...
 } ArmOp;
 
-#define check_cond(cond) if (!(cond)) { return data; }
+#define check_cond(cond) \
+	if (!(cond)) { \
+		return data; \
+	}
 
-static int get_mem_option(char *token) {
+static int get_mem_option (char *token) {
```
Comment (trufae) [inline libr/arch/p/arm/armass64.c:96]: function definitions don't have a space before the ( revert all those changes

File: libr/arch/p/arm/armass64.c [hunk 2]
```diff
@@ -84,9 +88,12 @@ typedef struct Opcode_t {
 	Operand operands[MAX_OPERANDS];
 } ArmOp;
 
-#define check_cond(cond) if (!(cond)) { return data; }
+#define check_cond(cond) \
```
Comment (trufae) [inline libr/arch/p/arm/armass64.c:91]: one line reads better with grep, also multiline defines should be avoided if necessary

File: libr/arch/p/arm/armass64.c [hunk 3]
```diff
@@ -1,5 +1,9 @@
 /* radare - LGPL - Copyright 2015-2023 - pancake */
 
+#include "r_types.h"
```
Comment (trufae) [inline libr/arch/p/arm/armass64.c:3]: all thse includes are already in by r_util.h

Comment (trufae) [inline libr/arch/p/arm/armass64.c:3]: define R_LOG_ORIGIN before any include

File: libr/arch/p/arm/armass64.c (hunk 1 above)
Comment (nathansrf) [reply libr/arch/p/arm/armass64.c:96]: Do you folks have a formatter that you run over the code, I believe these are from clang-d or something.

File: libr/arch/p/arm/armass64.c [hunk 4]
```diff
@@ -222,25 +228,48 @@ static inline ut32 encode1reg(ArmOp *op) {
...
+	R_LOG ("uimm6: %d", ui6);
+	ut32 data;
+
+	if (!(ui6 % 16)) {
+		data = ui6 / 16;
+	} else {
+		return 0;
+	}
+
+	R_LOG ("Encoded uimm6: %d", data);
```
Comment (trufae) [inline libr/arch/p/arm/armass64.c:264]: specify the log level using R_LOG_DEBUG for example

**Review by trufae — CHANGES_REQUESTED** [review]
looks good. despite some stylish changes here and there

File: libr/arch/p/arm/armass64.c (hunk 1 above)
Comment (nathansrf) [reply libr/arch/p/arm/armass64.c]: I ran the ./sys/clang-format-diff.py script over the code, hopefully this resolves these kind of issues.

File: libr/arch/p/arm/armass64.c (hunk 3 above)
Comment (nathansrf) [reply libr/arch/p/arm/armass64.c]: removed in latest commit

Comment (nathansrf) [reply libr/arch/p/arm/armass64.c]: Not sure I understand this one, I removed r_types.h but is there an alternative header I should be using instead?

File: libr/arch/p/arm/armass64.c (hunk 2 above)
Comment (nathansrf) [reply libr/arch/p/arm/armass64.c]: I have made this one-line using the GitHub web GUI, which hopefully works for now

**Comment by nathansrf** [comment]
I have tried to implement all of these changes, the only one that doesn't make much sense to me right now is the suggest R_LOG_ORIGIN change. Is this a different header file or macro I need to use?

**Comment by nathansrf** [comment]
I think the previous commit re-introduces a regression related to windows build support, so might need to remove it?

File: libr/arch/p/arm/armass64.c [hunk 5]
```diff
@@ -121,7 +124,7 @@ static int countTrailingZeros(ut64 x) {
 	return count;
 }
 
-R_UNUSED static bool has64reg(const char *str) {
+R_UNUSED static bool has64reg (const char *str) {
```
Comment (trufae) [inline libr/arch/p/arm/armass64.c:127]:
```suggestion
R_UNUSED static bool has64reg(const char *str) {
```

**Comment by trufae** [comment]
Merged! Thank you! Amazing pr and good tests here and there!

**Pull request merged**